### PR TITLE
First pull request

### PR DESCRIPTION
**Pull request by bparr**
Test.

### CODE REVIEW

File: agent.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 from menu_manager import characters
 import ctype_util as ct
 import pprint
+import pickle
```
Comment (bparr) [inline agent.py]: Usually want to have imports sorted alphabetically (https://google.github.io/styleguide/pyguide.html#Imports_formatting) but this file already doesn't follow this convention, so not needed here. New files should still be alphabetic though.

File: agent.py [hunk 2]
```diff
@@ -136,35 +139,63 @@ def act(self, state, pad):
...
-    # elif (self.frame_counter % 20) < 10 and ai_x > 15:
-    #   action = 3
-
-    # AI which side+A attacks the opponent when they're nearby
-    if (self.frame_counter % 30) < 20 and ai_x > cpu_x and ai_x-cpu_x < 20:
-      action = 8
-    elif (self.frame_counter % 30) < 20 and ai_x < cpu_x and cpu_x-ai_x < 20:
-      action = 9      
-
+    # # Jump three times then do nothing.
```
Comment (bparr) [inline agent.py]: Is there a reason to keep this code around in the codebase? Version control has it stored already. Are you planning on uncommenting anytime soon? If so, then this diff would be way smaller if used
"""
Everything in here is a comment.
"""

style comments.

File: agent.py [hunk 3]
```diff
@@ -136,35 +139,63 @@ def act(self, state, pad):
...
+      cur_L = self.memory.as_list()[-1].state.players[0].controller.button_L
+
+      if (not self.prev_L) and cur_L:
+        rl_model.toggle()
+
+      self.prev_L = cur_L
+
+      rl_model.update(cur_state, reward)
+      action = rl_model.act(cur_state)
+      print(action)
```
Comment (bparr) [inline agent.py]: Is this still needed? Maybe only pring if "verbose" is true? We are going to want to keep stdout fairly empty in the long run so it is easier to debug future code.

File: agent.py [hunk 4]
```diff
@@ -136,35 +139,63 @@ def act(self, state, pad):
...
+
+    if (self.frame_counter % 36000 == 0):
```
Comment (bparr) [inline agent.py]: Out of curiousity, was this able to complete in the 1/60 second time window we are working in? Or does it consitantly cause skipped frames?

File: cpu.py [hunk 5]
```diff
@@ -197,14 +204,15 @@ def spam(self, button):
...
+    def make_action(self, rl_model):
         # menu = Menu(self.state.menu)
         # print(menu)
+        # rl_model.update()
         if self.state.menu == Menu.Game.value:
             for pid, pad in zip(self.pids, self.pads):
                 agent = self.agents[pid]
                 if agent:
-                    agent.act(self.state, pad)
+                    agent.act(self.state, pad, rl_model)
```
Comment (bparr) [inline cpu.py]: After looking at this more closely, I think rl_model shouldn't be initialized in this file. Right now it's being created in this file and piped through two methods just to be passed to agent. Why not just initialize in agent.py constructor?

File: sarsa.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+		:param num_states: Number of the states in RL setup
+
+		:type num_actions: int
+		:param num_actions: Number of the actions in RL setup
+
+		:type learning_rate: float
+		:param learning_rate: Learning rate to use in algorithm
+
+		:type exploration: float
+		:param exploration: Probability of exploration during each action selection instance
```
Comment (bparr) [inline sarsa.py]: This style of documentation is new to me. Where is it from?

File: sarsa.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+		# Cumulative regard to keep track of progress
+		self.cum_reward = 0
+		self.cum_reward_list=[]
+
+		# Flag indicating whether or not to explore
+		self.explore_on = True
+
+		# Total number of updates so far
+		self.frames_trained = 0
```
Comment (bparr) [inline sarsa.py]: +1 for this constructor. Good job.

I always like constructors that are just "self.asdf = asdf" etc. It makes testing less of a headache. Yay.

File: sarsa.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+	def act(self, state):
```
Comment (bparr) [inline sarsa.py]: 2 spaces per tab vs. 4 spaces per tab?

I think Vlad uses 4 in his code, no?

File: sarsa.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+		Returns:
+		--------
+
+		:type action: int
+		:param action: Action to take
+
+		"""
+
+		if ((not self.explore_on) or np.random.uniform(0,1)) > self.exploration:
```
Comment (bparr) [inline sarsa.py]: Woah. This if statement is (bool or float) > float. Move the (not self.explore_on) so it's bool or (float > float)?

File: sarsa.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+		return reward
+
+	def update(self, cur_state, reward):
+		"""
+		Performs updates according to a specific algorithm
+
+		Arguments:
+		----------
+
+		:type cur_state: int
```
Comment (bparr) [inline sarsa.py]: You seem to like abreived_var_names. haha

File: agent.py [hunk 11]
```diff
@@ -136,35 +139,63 @@ def act(self, state, pad):
...
+
+      cur_L = self.memory.as_list()[-1].state.players[0].controller.button_L
+
+      if (not self.prev_L) and cur_L:
+        rl_model.toggle()
+
+      self.prev_L = cur_L
+
+      rl_model.update(cur_state, reward)
+      action = rl_model.act(cur_state)
```
Comment (bparr) [inline agent.py]: If I understand correctly, I think this should be named get_action. Also, I think coordinate_to_state and update should be merged in to this get_action method. It makes it easier for the caller of rl_model since it doesn't have to maintain either cur_state or reward. Just:

action = rl_model.get_action(x, y)

File: sarsa.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+		Returns:
+		--------
+
+		:type reward: int
+		:param reward: Reward based on the current state
+		"""
+
+		if np.sqrt(pow(x,2) + pow(y,2)) < 20:
+			# Near the center
+			reward = 1
```
Comment (bparr) [inline sarsa.py]: Use np.linalg.norm([x, y])? Also, this could be a bit shorter if just return once reward is known. Though with my previous comment of merging this into get_action, the "return" shorter code doesn't make sense anymore. Eh.

```
if np.linalg.norm([x, y]) < 20:
  return 1  # Near the center
if x < -60 or x > 60:
  return -1  # Off-state
return 0
```

File: sarsa.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+		if ((not self.explore_on) or np.random.uniform(0,1)) > self.exploration:
+			action = np.argmax(self.q[state])
+		else:
+			action = np.random.choice(range(self.num_actions))
+
+		self.prev_action = action
+
+		return action
+
+	def toggle(self):
```
Comment (bparr) [inline sarsa.py]: Rename to toggleExploration.

File: sarsa.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+		:type cur_state: int
+		:param cur_state: The current state of the agent
+
+		:type reward: int
+		:param reward: Reward received based on the previous state and action
+
+		"""
+		best_action = np.argmax(self.q[cur_state])
+
+		if self.model == 'qlearning':
```
Comment (bparr) [inline sarsa.py]: Are other models supported? What happens when not equal to 'qlearning'?

File: sarsa.py [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+			# Q-Learning : https://en.wikipedia.org/wiki/Q-learning
+			self.q[self.prev_state][self.prev_action] += (self.learning_rate * (
+											             reward + self.discount*self.q[cur_state][best_action]
```
Comment (bparr) [inline sarsa.py]: Something is off with this formatting. Please fix.

File: sarsa.py [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+											             - self.q[self.prev_state][self.prev_action]))
+
+		self.prev_state = cur_state
+		self.cum_reward = 0.0001*reward + 0.9999*self.cum_reward
```
Comment (bparr) [inline sarsa.py]: Spaces around * and after the "if" below.

File: sarsa.py [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+		:param y: y-coordinate of agent
+
+		Returns:
+		--------
+
+		:type state: int
+		:param state: State of the agent
+
+		"""
+		state_x = math.floor(x/10)
```
Comment (bparr) [inline sarsa.py]: I believe the following code is equivalent to this code. Also, in general hardcoded constants like this scare me, but this is a special case where I think it's fine (since this code's use is limited to step 3.0 and maybe 3.1).

```
state_x = np.clip(np.floor(x / 10), -7, 6) + 7
state_y = np.clip(np.floor(y / 10), -1, 8) + 1
return state_x * 10 + state_y

```

File: sarsa.py [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+
+		if state_y < -1:
+			state_y = -1
+
+		if state_y > 8:
+			state_y = 8
+
+		state_y = state_y +1
+
+		return state_x*10 + state_y
```
Comment (bparr) [inline sarsa.py]: Is Q-Learning able to be given both state_x and state_y instead of a combined version? Eventually our RL will need to be able to see multiple variables.

File: sarsa.py [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+
+		:type learning_rate: float
+		:param learning_rate: Learning rate to use in algorithm
+
+		:type exploration: float
+		:param exploration: Probability of exploration during each action selection instance
+
+		"""
+
+		self.num_states = num_states
```
Comment (bparr) [inline sarsa.py]: Please add '_' to these class variables to mark them as protected (https://google.github.io/styleguide/pyguide.html?showone=Naming#Naming).

File: agent.py (hunk 1 above)
Comment (deepakdilipkumar) [inline agent.py]: Will keep that in mind

File: agent.py (hunk 2 above)
Comment (deepakdilipkumar) [inline agent.py]: No, you're right. It can be removed

File: agent.py (hunk 11 above)
Comment (deepakdilipkumar) [inline agent.py]: I've been meaning to make some changes similar to that myself, because my current flow makes it a pain to implement stuff like sarsa. I'll make those changes in my next commit (probably not today though).

File: agent.py (hunk 3 above)
Comment (deepakdilipkumar) [inline agent.py]: Yes, and action specifically can definitely be removed from print statements.

File: agent.py (hunk 4 above)
Comment (deepakdilipkumar) [inline agent.py]: Didn't exactly get what you mean

File: sarsa.py (hunk 8 above)
Comment (deepakdilipkumar) [inline sarsa.py]: I'm mostly using 4 as well. Did you find some place where it's 2 places per tab? These lines in sarsa.py are already 4 spaces I think.

File: sarsa.py (hunk 6 above)
Comment (deepakdilipkumar) [inline sarsa.py]: http://deeplearning.net/tutorial/lenet.html

File: sarsa.py (hunk 9 above)
Comment (deepakdilipkumar) [inline sarsa.py]: Good catch. Changed.

File: sarsa.py (hunk 13 above)
Comment (deepakdilipkumar) [inline sarsa.py]: Done

File: sarsa.py (hunk 12 above)
Comment (deepakdilipkumar) [inline sarsa.py]: Higher priority to merging into one function, like you said. I'll shorten it if possible while doing that. I might still be calling the reward() function from within get_action.

File: sarsa.py (hunk 10 above)
Comment (deepakdilipkumar) [inline sarsa.py]: current_state is preferable?

File: sarsa.py (hunk 14 above)
Comment (deepakdilipkumar) [inline sarsa.py]: It DIES

Comment (deepakdilipkumar) [inline sarsa.py]: jk it does anything. Other models will be supported in the future. I can add a warning or default it to q-learning or something in case self.model not equal something we know.

File: sarsa.py (hunk 15 above)
Comment (deepakdilipkumar) [inline sarsa.py]: Huh. Looks normal-ish locally. I'll check.

File: sarsa.py (hunk 18 above)
Comment (deepakdilipkumar) [inline sarsa.py]: Not sure. I'll need to read up more.

File: sarsa.py (hunk 17 above)
Comment (deepakdilipkumar) [inline sarsa.py]: Nice! Thanks.
